### ISSUE TITLE
Fedora hardlink vs Debian hardlink

### ISSUE THREAD

**Issue by julian-klode** [karelzak/util-linux#808]
It seems you have merged the Fedora version of hardlink into util-linux. This is unfortunate, as Debian, Ubuntu, Arch, and others are shipping a different hardlink, developed my me, that has a lot more features:

```
Usage: ./hardlink [options] directory|file ...
Options:
  -V, --version         show program's version number and exit
  -h, --help            show this help message and exit
  -v, --verbose         Increase verbosity (repeat for more verbosity)
  -n, --dry-run         Modify nothing, just print what would happen
  -f, --respect-name    Filenames have to be identical
  -p, --ignore-mode     Ignore changes of file mode
  -o, --ignore-owner    Ignore owner changes
  -t, --ignore-time     Ignore timestamps (when testing for equality)
  -X, --respect-xattrs  Respect extended attributes
  -m, --maximize        Maximize the hardlink count, remove the file with
                        lowest hardlink cout
  -M, --minimize        Reverse the meaning of -m
  -O, --keep-oldest     Keep the oldest file of multiple equal files
                        (lower precedence than minimize/maximize)
  -x REGEXP, --exclude=REGEXP
                        Regular expression to exclude files
  -i REGEXP, --include=REGEXP
                        Regular expression to include files/dirs
  -s <num>[K,M,G], --minimum-size=<num>[K,M,G]
                        Minimum size for files. Optional suffix
                        allows for using KiB, MiB, or GiB

Compatibility options to Jakub Jelinek's hardlink:
  -c                    Compare only file contents, same as -pot
```

Switching to this version of hardlink would break things for a lot of users. Apart from the meaning of `-f`, mine is a superset option-wise.

I'd argue given that my version is a superset of this version, we should probably consolidate on this one. I think I mentioned this to Jakub back in the day, but I'm not sure.

**Comment by julian-klode** [comment]
FWIW, my hardlink is maintained at https://salsa.debian.org/jak/hardlink at the moment.

**Comment by julian-klode** [comment]
As explained in Arch bug https://bugs.archlinux.org/task/62896, switching to the one you're shipping would break things, as people rely on the additional options.

If we do merge, we'd have to figure out what to do about `-f`. Probably the only sane thing is to drop support for it, and only keep the long names, so scripts would work correctly either way, or fail reliably.

**Comment by karelzak** [comment]
Ah, nice to see that someone is complain about the merge *now*. We have discussion about it on mailing list months before the merge, we have -rc1 and -rc2.. :-( Anyway, --disable-hardlink is elegant way how to avoid possible problem, nobody is forced to use it.

The ideal solution would be to merge also your version into util-linux and create one final hardlink version for all distros. You're right that -f is annoying, but I think it something we can resolve (probably don't use -f at all in next versions). If you agree, we can:

- inform about this issue on ulti-linux mailing list
- temporary disable util-linux hardlink on Debian (just --disable-hardlink)
- analyse possible differences between both versions
- prepare patches to merge both version
- release util-linux v2.35 (in next 4-6 months) with final hardlink

What do you think?

**Comment by karelzak** [comment]
Note that this is ideal opportunity to consolidate things.

IMHO the option -f  on Fedora is probably rarely used use-case, maybe we can ignore this issue and use -f (=basename) in your way only. It will very probably break backward compatibility less widely than remove -f at all from all versions.

**Comment by julian-klode** [comment]
Oh, I just got notified yesterday by the friendly Arch folks that there's an overlap. Debian's in a freeze and Ubuntu has not updated binutils either, so we have not noticed this on our own.

I just looked at the mailing list and found https://www.spinics.net/lists/util-linux-ng/msg15273.html where it was mentioned that we have the different version in Debian, albeit the author was a bit confused as it used to be Python but switched to C in 0.3. Sadly, he did not CC me.

I'm not sure it's clear that they are distinct code bases, only sharing the option names. Jakub Jelinek's hardlink.c was rewritten in Python as hardlink.py, which I then rewrote because their code was hard to follow. That was then rewritten again in C. This all happened before `-f` was added to hardlink.c, hence the divergence. It's also MIT licensed at the moment.

I'm fine with merging it into util-linux, as it will get more people to work on it. I guess it's a bit sad for the BSD folks, but they can fork their own one if they want something more pure. I'm much more concerned by having two hardlink variants that we'd have to keep in sync option-name wise.

**Comment by karelzak** [comment]
It's possible to build subset of util-linux on FreeBSD, Hurd or OSX ... it is not strictly about Linux (but Linux is our primary target). So, BSD folks probably do not have to fork it.

Frankly, your code is more readable for me than Jakub's original implementation :-) 

The small difference (but without impact to the result) is that util-linux version counts checksum from begin of the file (64 bytes) to avoid unnecessary nodes in memory, but it means extra open+read. Your version is happy with struct stat from nftw() only. I guess your version probably needs more memory, but less syscalls. For me the checksum is unnecessary optimization in these days.

MIT license is probably no problem, although GPLv2+ would be probably better to minimize licences variability in util-linux.

That's cool you agree with the merge. I'll try to prepare a new util-linux branch with your hardlink, add some patches (use our xmalloc(), xstrdup(), strtosize(), size_to_human_string() for {K,M,G...}iB, etc),  merge man pages. And than I'll ask you for review.

**Comment by eli-schwartz** [comment]
I too wish someone had noticed this beforehand -- but on our (Arch Linux) end, we only noticed this when util-linux was released to our testing repos and someone reported a bug that they could not install it due to namespace conflicts: https://bugs.archlinux.org/task/62896

**Comment by julian-klode** [comment]
I'm fine with relicensing to GPL-2+, but we'd also need consent from Intel for their xattr contribution, but I think that's about it, as the other contributions were single character typo fixes.

On the other hand, it does not matter much given that the binary becomes GPL anyway.

**Comment by julian-klode** [comment]
And, yes, I tried very hard to make it readable, and well-documented, and the code is also auto-formatted with indent I think to make sure that's consistent :-)

**Comment by Code7R** [comment]
@julian-klode Please, if you deprecate a project or branch, make sure that it's deprecation is visible if to people who use the information from APT to clone the repo!

I just spent some time to implement a little feature and only then discovered this activity here :-(

Please consider merging this stuff anyway: https://salsa.debian.org/blade/hardlink/commits/feature/custom_read_buffer_size

**Comment by NoelJB** [comment]
What is the status of the merger?  I just looked at the master branch for util-linux, and it appears that the obsolete version is still present, despite no further conversation since June of last year.

**Comment by karelzak** [comment]
No progress. I'm busy with some other stuff (right now with libmount and new kernel mount API)...

Any volunteers? ;-)

**Comment by NoelJB** [comment]
What needs to be done, other than replacing the code?  Just this?  "Prepare a new util-linux branch with [Klode's]] hardlink, add some patches (use [Zak's] xmalloc(), xstrdup(), strtosize(), size_to_human_string() for {K,M,G...}iB, etc), merge man pages."

**Comment by JohnVillalovos** [comment]
FYI: Likely the version at https://salsa.debian.org/jak/hardlink is much faster than the one here. I see the Debian one is based on my Python version which was much faster than the original hardlink.c. I imagine the C version is faster than my Python version https://github.com/JohnVillalovos/hardlinkpy.

Unsure if the one here has been updated to improve the algorithm used, which was the cause of the slowness.

**Comment by ChrisEineke** [comment]
@Code7R Same here. Just started working on a small feature and came upon this discussion.

Is there still interest in getting this done? I wouldn't mind taking a stab at it.

**Comment by karelzak** [comment]
OK, finally I have time to work on hardlink. Yes, there is interest in getting this done. 

I'm going to create the branch with @julian-klode version.

**Comment by karelzak** [comment]
Just for the record, see topic/hardlink branch https://github.com/karelzak/util-linux/commits/topic/hardlink -- I'm going to work on it in the next days.

**Comment by karelzak** [comment]
Debian version merged to the master branch. 

@julian-klode thanks for your code!

**Comment by julian-klode** [comment]
Hooray! Thanks for merging.

**Comment by eli-schwartz** [comment]
Yay, thank you so much for both of your hard work. :D
